genwinsyscallproto/genwinsyscalldef.go

```diff
@@ -0,0 +1,201 @@
+// required. (mkwinsyscall has defaults that may be acceptable.) It is recommended to read the
+// DllImport pseudo-attribute (§II.21.2.1) to determine this value.
+func WriteMethod(w io.StringWriter, metadata *winmd.Metadata, method *winmd.MethodDef, moduleName, goName string) error {
+	w.WriteString("//sys\t")
+	w.WriteString(goName)
+	w.WriteString("(")
+
+	for paramRowIndex := method.ParamList.Start; paramRowIndex < method.ParamList.End; paramRowIndex++ {
+		param, err := metadata.Tables.Param.Record(paramRowIndex)
+		if err != nil {
+			return fmt.Errorf("failed to read param row %v defined by method %v: %w", paramRowIndex, method.Name, err)
+		}
+		if param.Sequence == 0 {
+			// Sequence 0 is ".param". A row with Sequence 0 is often included before other rows,
+			// but Sequence 0 doesn't contain any data in cases seen so far.
+			if param.Flags == 0 && param.Name.String() == "" {
+				continue
+			}
+			// TODO: Add support for assemblies that do have data in Sequence 0.
+			return fmt.Errorf("unsupported method: expected param row with sequence 0 to be empty, but: %#v", param)
+		}
+
+		// Param data is in two places: a param table row and the signature. Index into the
+		// signature param slice by converting the 1-based sequence value. Note that we assume
+		// ascending Sequence values for proper formatting: this is stated to be true in the
+		// informational section of §II.22.33, point 4.
+		// TODO: Handle gaps in Sequence values? §II.22.33 information point 5.
+		// Gaps have not been seen in Windows.Win32.winmd. The meaning of a gap is not yet known.
+		i := param.Sequence - 1
+		if i > 0 {
+			w.WriteString(", ")
+		}
+		writeParam(w, param)
+		w.WriteString(" ")
+
+		if int(i) >= len(method.Signature.Param) {
+			return fmt.Errorf("param record Sequence value %v is out of range of parsed signature params, length %v", i, len(method.Signature.Param))
+		}
+		if err := writeType(w, metadata, method.Signature.Param[i].Type); err != nil {
+			return fmt.Errorf("failed to interpret type of param %v of method %v: %w", i, method.Name, err)
+		}
+	}
+	w.WriteString(")")
+
+	// Write return value, if one exists.
+	if method.Signature.RetType.Kind != winmd.RetTypeKind_Void {
+		w.WriteString(" (")
+		if err := writeType(w, metadata, method.Signature.RetType.Type); err != nil {
+			return err
+		}
+		w.WriteString(")")
+	}
+
+	// Write syscall name and module if non-defaults are needed.
+	if goName != method.Name.String() || moduleName != "" {
+		w.WriteString(" = ")
+		if moduleName != "" {
+			w.WriteString(moduleName)
+			w.WriteString(".")
+		}
+		w.WriteString(method.Name.String())
+	}
+	return nil
+}
+
+func writeParam(w io.StringWriter, p *winmd.Param) {
+	// TODO: Remove these? Or use them to help determine how to translate return types for syscalls.
+	// Include comments for param attributes to make them visible for diagnosis purposes.
+	if p.Flags&flags.ParamAttributes_In != 0 {
```

Comment:
nit: the output would look cleaner with this format: `/*in,optional,hasdefault*/`. This is easy to implement using something like:

```go
var attrs []string
if foo {
  attrs = append(attrs, "foo")
}
w.WriteString("/*" + strings.Join(attrs, ",") + "*/")
```

Reply:
Not sure yet whether mkwinsyscall will handle these comments, so actually writing something here might just be temporary, but if it ends up working, agreed that this format is significantly easier to read.

Maybe a diag mode for the tool/library would make sense, which would write additional lines above each `//sys` line describing why it made the decisions it did for that particular method. Or maybe it's overkill vs. simply debugging the tool. Hmm.

Reply:
mkwinsyscall won't handle those comments, but good to have somewhere for troubleshooting.

Reply:
I'm leaving the info out for this PR, but filed https://github.com/microsoft/go-winmd/issues/25 to get more diagnosis data into the tool. (A verbose logging mode might be fine--putting the info into the file itself doesn't seem like the most important thing here.)